lib/net/http.rb

```diff
@@ -1654,14 +1654,22 @@ def connect
       end
 
       debug "opening connection to #{conn_addr}:#{conn_port}..."
-      s = Timeout.timeout(@open_timeout, Net::OpenTimeout) {
-        begin
-          TCPSocket.open(conn_addr, conn_port, @local_host, @local_port)
-        rescue => e
-          raise e, "Failed to open TCP connection to " +
-            "#{conn_addr}:#{conn_port} (#{e.message})"
+      begin
+        s = begin
+          # Use built-in timeout in TCPSocket.open if available
+          TCPSocket.open(conn_addr, conn_port, @local_host, @local_port, open_timeout: @open_timeout)
+        rescue ArgumentError => e
```

Comment:
I couldn't come up with a better way to detect the absence of `open_timeout:` option. Is this a acceptable solution?

TCPSocket.open is basically `(**args)` from the perspective of Ruby, so Method#parameters wasn't an option:

```
irb(main):001> require 'socket'
=> true
irb(main):002> TCPSocket.method(:open).parameters
=> [[:rest]]
```

Reply:
Actually it's not only very old Rubies, but even 3.4 raises on a `TCPSocket.open` call with `open_timeout`. It is true that the situation is different in Ruby 2.x, where keyword arguments were not a argument of its own kind (workaround in 09bf573dd5bd7111b56c57918c8a3a84b068309b).

Reply:
Oh, I thought your question was specifically for the Ruby 2.x workaround, since you need to parse a very generic error message, but I see how the detection for Ruby 3.4 is also brittle since it also involves parsing the error message, even if more specific. Unfortunately, I don't know of a better way, but I'd say this way is acceptable.

Reply:
Considering that every single new HTTP request is going to go through this raise/rescue flow on ruby < 3.5, and that exceptions are kind of expensive, personally I think a better option is just to have test for `RUBY_VERSION.to_f < 3.5` directly to see if we use the old `Timeout.timeout`. I get that testing version numbers directly is a bit distasteful, but when there's actual performance issues and brittleness issues on the line... I'd say its merited

Reply:
I've took @eregon's idea. Pushed in 06d982f3a178a65777332ed402d5a80c4b3216da. (thanks: @niku)

Reply:
Yes, that’s exactly right. Using class variables would unfortunately kill Ractors.

Reply:
> `resolv_timeout` is extremely complex to handle, CRuby has had many iterations of it (some of them buggy, leaking, segfaulting and other issues)

Can't resist to link https://bugs.ruby-lang.org/issues/21790 which is another case of "it's very messy to do getaddrinfo with timeout i.e. resolv_timeout".